### PR TITLE
enable data type propagation

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98065

Enable data type propagation in schedule node level.
Propagation policy:
(1) ops with dtype args [constant, load, rand, randn] -> direct use dtype as node dtype
(2) ops semantics decide output dtype -> using output dtype
All `override_return_dtype` in https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py.
(3) other ops: perform promote on input nodes dtype. ADD(BF16, FP32) -> FP32 output.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
Have you validated this PR with E2E models?

**Comment by zhuhaozhe** [comment]
> Have you validated this PR with E2E models?

I verified on a tiny model and some other ut to observe the dtype attached.

**Review by jansel — CHANGES_REQUESTED** [review]
I don't think this needs to be specific to cpp codegen.  Let's move it to its own file.

**Comment by zhuhaozhe** [comment]
> I don't think this needs to be specific to cpp codegen. Let's move it to its own file.

Thanks, I removed it to `common.py`and invoke it in `cpp` path.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
...
+        opt_ctx.dtype = dtype
+        msg = f"for node.target = {_node.target}, dtype is propagated to {opt_ctx.dtype}, "
+        input_msg = "inputs dtypes: "
+        for n in input_nodes:
+            input_msg += (
+                f"input {n.name}.dtype = {n.meta[OptimizationContext.key].dtype}"
+            )
+        data_type_logger(msg + input_msg)
+        _node.meta[OptimizationContext.key] = opt_ctx
+        opt_ctx.dtype_propogated = True
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:130]: Is this a duplicate of `opt_ctx.dtype`? Can we check just check if `opt_ctx.dtype` is `None` or not?

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
...
+            "eq",
+            "ne",
+            "bitwise_not",
+            "bitwise_or",
+            "bitwise_left_shift",
+            "bitwise_right_shift",
+        ]
+        ops_with_dtype_arg = ["constant", "to_dtype", "rand", "randn"]
+        reduction_to_bool = ["any"]
+        reduction_to_int64 = ["argmin", "argmax"]
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:71]: Simpler to build a dict from reduction type to dtype?

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
...
+            "logical_not",
+            "signbit",
+            "le",
+            "lt",
+            "ge",
+            "gt",
+            "eq",
+            "ne",
+            "bitwise_not",
+            "bitwise_or",
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:65]: I suppose the result can be either bool or integer? Why it has to be bool?
https://pytorch.org/docs/stable/generated/torch.bitwise_or.html

File: torch/_inductor/codegen/common.py [hunk 4]
```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
     return [*index, sympy_dot(index_vars, FlexibleLayout.contiguous_strides(sizes))]
 
 
+def _data_type_propagation(sub_graph: torch.fx.Graph):
+    def propagate_node(node: torch.fx.Node):
+        _node: torch.fx.Node = node
+        if _node.target == "ops":
+            return False
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:51]: Check target in `ops_without_dtype` instead? Then, you don't have to explicitly check `ops_without_dtype` later?

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py:130]: Yes, I removed this.

File: torch/_inductor/codegen/common.py (hunk 2 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py:71]: Done

File: torch/_inductor/codegen/common.py (hunk 4 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py]: Yes, we can check `target in ops_without_dtype` here, but we still need to explicitly check ops_without_dtype later while propagate dtype from input nodes. 
```
        input_nodes = [
            n
            for n in inputs
            if isinstance(n, torch.fx.node.Node) and n.target not in ops_without_dtype
        ]
```
While `node.inputs.target = ['ops' ,'get_index', 'add'],` we need remove `ops` and `get_index` and only use `add` to decide the `node.dtype`

File: torch/_inductor/codegen/common.py (hunk 3 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py:63]: Thanks for reminding here, I removed `bitwise_not, bitwise_or, bitwise_left_shift, bitwise_right_shift` here and also add UT for this.

**Review by jgong5 — APPROVED** [review]

**Comment by EikanWang** [comment]
@zhuhaozhe , please check the failed cases.

**Pull request closed without merging**